### ISSUE TITLE
broken naming/rendering of aliased subcharts

### ISSUE THREAD

**Issue by patrikbeno** [helm/helm#11484]
Experimenting with 
- multiple levels of subcharts
- aliased dependencies: same subchart included multiple times with different alias

Findings:
- same subchart, multiple aliases work OK
- unless such subchart has other dependencies
  - naming is broken (mixed aliases/chart names, i.e. 1st subchart is rendered with aliases, second uses chart name instead of alias) 
  - see [bug1.zip](https://github.com/helm/helm/files/9891165/bug1.zip)
- and if such subchart does the similar thing (same chart aliased twice): 
  - naming is broken 
  - not all subcharts are fully rendered (missing files).
  - see [bug2.zip](https://github.com/helm/helm/files/9891167/bug2.zip)
  - i.e. first subchart (l2a) is rendered OK, second (l2b) has broken naming (chart name instead of alias), and as a consequence, its template (level3) is rendered only once (instead of twice)

Consider this:

```
---
apiVersion: v2
name: level1
version: 0.1.0
dependencies:
- name: level2
  version: "*"
  repository: file://../level2
  alias: l2a
- name: level2
  version: "*"
  repository: file://../level2
  alias: l2b

---
apiVersion: v2
name: level2
version: 0.1.0
dependencies:
- name: level3
  version: "*"
  repository: file://../level3
  alias: l3a
- name: level3
  version: "*"
  repository: file://../level3
  alias: l3b
```

Output of `helm version`:
3.10.1

**Comment by joejulian** [comment]
Not a duplicate of #9214

**Comment by joejulian** [comment]
I'd walk this with a debugger and see if you can find the problem. https://github.com/helm/helm/blob/7ed9d16dc764a5b94b378a7e217865efaa0d9ac8/pkg/chartutil/dependencies.go#L93-L115.

**Comment by patrikbeno** [comment]
still an issue

**Comment by patrikbeno** [comment]
not stale, reopen pls

**Comment by patrikbeno** [comment]
I tested #12276 against the test cases reported in this issue, and it seems working.

**Comment by aabouzaid** [comment]
This sentence is not correct anymore (Helm CLI v3.16.1)

> same subchart, multiple aliases work OK

I'm unsure if that's a new issue or just because this issue was a bit old (reported in 2022).

Something happens in the packaging step where the chart name (not the alias) is used to create the deps charts.

It doesn't matter if the sub-chart has its own dependency or not, it always happens if there are 2 sub-chart from the same Helm repo even with different aliases.

---

When I package this chart:

```
apiVersion: v2
name: my-chart
version: 1.0.0
appVersion: 1.0.0
dependencies:
  - name: postgresql
    alias: psql-14
    repository: https://charts.bitnami.com/bitnami
    version: 14.x.x
  - name: postgresql
    alias: psql-13
    repository: https://charts.bitnami.com/bitnami
    version: 13.x.x
```

It creates 1 PostgreSQL Frankenstein package called postgres which has some files from the Bitnami chart postgresql/13.x.x and postgresql/14.x.x!

![single-frankenstein-deps-chart](https://github.com/user-attachments/assets/f6a5ae4f-ce81-4d7d-863f-1702c4a33620)
